std/machines/hash/poseidon_bb.asm

```diff
@@ -0,0 +1,254 @@
+
+    let operation_id;
+
+    // Number of field elements in the state
+    let STATE_SIZE: int = 16;
+    // Number of output elements
+    // (TODO: to use the sponge construct, the entire state should be output)
+    let OUTPUT_SIZE: int = 8;
+    // Number of full rounds (half of them before and half of them after the partial rounds)
+    let FULL_ROUNDS: int = 8;
+    // Number of partial rounds
+    // For 128-bit security, Poseidon2 requires 13 partial rounds. But we need more here
+    // To have enough rows for memory read / write.
+    let PARTIAL_ROUNDS: int = 16;
+    let ROWS_PER_HASH = FULL_ROUNDS + PARTIAL_ROUNDS + 1;
+
+
+    // ------------- Begin memory read / write ---------------
+
+    // Get an intermediate column that indicates that we're in an
+    // actual block, not a default block. Its value is constant
+    // within the block.
+    let used = array::sum(sel);
+    array::map(sel, |s| unchanged_until(s, LAST));
+    std::utils::force_bool(used);
+
+    // Repeat the input state in the whole block
+    let input: col[STATE_SIZE];
+    array::map(input, |c| unchanged_until(c, LAST));
+    array::zip(input, state, |i, s| CLK[0] * (i - s) = 0);
+
+    // Repeat the time step and input / output address in the whole block
+    let time_step;
+    let input_addr_high;
+    let input_addr_low;
+    let output_addr_high;
+    let output_addr_low;
+    unchanged_until(time_step, LAST);
+    unchanged_until(input_addr_high, LAST);
+    unchanged_until(input_addr_low, LAST);
+    unchanged_until(output_addr_high, LAST);
+    unchanged_until(output_addr_low, LAST);
+
+    // Increment the address in every row but the one it is set.
+    let is_addr_set = CLK[0] + CLK[STATE_SIZE];
+    let high_addr;
+    let low_addr;
+
+    // Set the initial address to read from
+    CLK[0] * (high_addr - input_addr_high) = 0;
+    CLK[0] * (low_addr - input_addr_low) = 0;
+
+    // Set the initial address to write to
+    CLK[STATE_SIZE] * (high_addr - output_addr_high) = 0;
+    CLK[STATE_SIZE] * (low_addr - output_addr_low) = 0;
+
+    // How far away from overflowing the low limb is:
+    let low_diff = low_addr - (0x10000 - 4);
+    // Is one if low limb is about to overflow:
+    let low_overflow;
+    low_overflow = 1 - low_diff_inv * low_diff;
+    // Helper to allow low_overflow to be boolean
+    let low_diff_inv;
+    // Ensures that (low_diff_inv * low_diff) is boolean,
+    // and that low_diff_inv is not 0 when low_diff is not 0:
+    (low_diff_inv * low_diff - 1) * low_diff = 0;
+
+    // Increment the low limb if address is not being set:
+    (1 - is_addr_set') * (
+        // If low limb is about to overflow, next value must be 0:
+        low_overflow * low_addr' +
+        // Otherwise, next value is current plus 4:
+        (1 - low_overflow) * (low_addr + 4 - low_addr')
+    ) = 0;
+    // Set high limb, incremented if low overflowed:
+    (1 - is_addr_set') * (high_addr' - high_addr - low_overflow) = 0;
```

Comment:
Does this block assume addresses to be 4-aligned?

Reply:
Yes. Starting from this line:
```
    let low_diff = low_addr - (0x10000 - 4);
```

 My hope is that address alignment is enforced somewhere else (in memory machine?). I know it is enforced for poseidon_gl when the call comes from RISC-V.

Reply:
yea I think it makes sense